src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+            // No need for NoSynchronizationContextScope.Enter() since
+            // NpgsqlConnection.OpenAsync() does it itself and it's the
+            // first/only async method call.
+            using var openState = EnsureAndSetState(ReplicationConnectionState.Closed, ReplicationConnectionState.Opening);
+            await _npgsqlConnection.OpenAsync(cancellationToken);
+
+            SetTimeouts(CommandTimeout, CommandTimeout);
+
+            openState.ChangeResetState(ReplicationConnectionState.Idle);
+        }
+
+        /// <summary>
+        /// Closes the replication connection and performs tasks associated
+        /// with freeing, releasing, or resetting its unmanaged resources asynchronously.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous dispose operation.</returns>
+        public ValueTask DisposeAsync()
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return DisposeAsyncInternal();
+        }
+
+        async ValueTask DisposeAsyncInternal()
+        {
+            var currentState =
+                (ReplicationConnectionState)Interlocked.Exchange(ref _state, (int)ReplicationConnectionState.Disposed);
+            if (currentState == ReplicationConnectionState.Disposed)
+                return;
+
+            // The expected sequence is probably to first cancel the streaming replication and then dispose
+            // the connection, but if DisposeAsync() is called while the connection is streaming, we
+            // do the cancellation ourselves to bring it down in a predictable way.
+            if (currentState == ReplicationConnectionState.Streaming)
+                Cancel();
+
+#if NETSTANDARD2_0
+            _sendFeedbackTimer.Dispose();
+            _requestFeedbackTimer.Dispose();
+#else
+            await _sendFeedbackTimer.DisposeAsync();
+            await _requestFeedbackTimer.DisposeAsync();
+#endif
+            _feedbackSemaphore.Dispose();
+            await _npgsqlConnection.Close(async:true);
+        }
+
+        #endregion Open / Dispose
+
+        #region Replication methods
+
+        /// <summary>
+        /// Requests the server to identify itself.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>
+        /// A <see cref="NpgsqlReplicationSystemIdentification"/> containing information
+        /// about the system we are connected to.
+        /// </returns>
+        public Task<NpgsqlReplicationSystemIdentification> IdentifySystem(CancellationToken cancellationToken = default)
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return IdentifySystemInternal(cancellationToken);
+        }
+
+        async Task<NpgsqlReplicationSystemIdentification> IdentifySystemInternal(CancellationToken cancellationToken = default)
```

Comment:
nit: can move this into the above as a local method (all of the others as well)

Reply:
Done in a39b6a6f8cc441601983d297ddf6c28e5f9141a1 and 4825c127551f4c81b69bdbc1a684a6ba109c7976